cmd/metacache-server-pool.go

```diff
@@ -95,156 +94,278 @@ func (z *erasureServerPools) listPath(ctx context.Context, o listPathOptions) (e
 	}
 
 	// Decode and get the optional list id from the marker.
-	o.Marker, o.ID = parseMarker(o.Marker)
-	o.Create = o.ID == ""
-	if o.ID == "" {
-		o.ID = mustGetUUID()
-	}
+	o.parseMarker()
 	o.BaseDir = baseDirFromPrefix(o.Prefix)
-	if o.discardResult {
-		// Override for single object.
-		o.BaseDir = o.Prefix
-	}
-
-	// For very small recursive listings, don't same cache.
-	// Attempts to avoid expensive listings to run for a long
-	// while when clients aren't interested in results.
-	// If the client DOES resume the listing a full cache
-	// will be generated due to the marker without ID and this check failing.
-	if o.Limit < 10 && o.Marker == "" && o.Create && o.Recursive {
-		o.discardResult = true
-		o.Transient = true
+	o.Transient = o.Transient || isReservedOrInvalidBucket(o.Bucket, false)
+	if o.Transient {
+		o.Create = false
 	}
 
-	var cache metacache
+	// We have 2 cases:
+	// 1) Cold listing, just list.
+	// 2) Returning, but with no id. Start async listing.
+	// 3) Returning, with ID, stream from list.
+	//
 	// If we don't have a list id we must ask the server if it has a cache or create a new.
-	if o.Create {
-		o.CurrentCycle = intDataUpdateTracker.current()
-		o.OldestCycle = globalNotificationSys.findEarliestCleanBloomFilter(ctx, path.Join(o.Bucket, o.BaseDir))
-		var cache metacache
+	if o.ID != "" && !o.Transient {
+		// Create or ping with handout...
 		rpc := globalNotificationSys.restClientFromHash(o.Bucket)
-		if isReservedOrInvalidBucket(o.Bucket, false) {
-			rpc = nil
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		c, err := rpc.GetMetacacheListing(ctx, *o)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				// Context is canceled, return at once.
```

Comment:
If you change this in this manner `rpc` is nil when its a local erasure setup, if you do not use `localCacheMgr` then  local erasure coded setups won't work - as you can see `mint` tests are again failing. 

This is the fix that I shared in a previous comment @klauspost